File: 0022-generate-parentheses/0022-generate-parentheses.cpp
class Solution {
    void solve(string op ,int open, int close, vector<string>& ans)
    {
        if(open==0 && close==0)
        {
            ans.push_back(op);
            return;
        }
        
        if(open == close)
        {
            string op1 = op;
            op1.push_back('(');
            solve(op1, open-1, close, ans);
        }
        else if(open==0)
        {
            string op1 = op;
            op1.push_back(')');
            solve(op1, open, close-1, ans);
        }
        // else if(close == 0){
        //     string op1 = op;
        //     op1.push_back('(');
        //     solve(op1, open-1, close, ans);
        // }
        else
        {
            string op1=op;
            string op2=op;
            op1.push_back('(');
            op2.push_back(')');
            solve(op1, open-1, close, ans);
            solve(op2, open, close-1, ans);
        }
    }
public:
    vector<string> generateParenthesis(int n) {
        string op="";
        int open=n , close=n;
        vector<string> ans;
        solve(op, open, close, ans);
        return ans;
    }
};

File: 0022-generate-parentheses/0022-generate-parentheses.java
class Solution {
    public List<String> generateParenthesis(int n) {
        List<String> ans = new ArrayList<>(); // Initialize the list
        generate(ans, n, 0, 0, ""); // Start the recursive function
        return ans;
    }

    public void generate(List<String> ans, int n, int open, int close, String cur_str) {
        // If the current string has the required length, add it to the result
        if (cur_str.length() == 2 * n) {
            ans.add(cur_str);
            return;
        }

        // If the number of open brackets is less than n, add an open bracket
        if (open < n) {
            generate(ans, n, open + 1, close, cur_str + "(");
        }

        // If the number of close brackets is less than open brackets, add a close bracket
        if (close < open) {
            generate(ans, n, open, close + 1, cur_str + ")");
        }
    }
}